### PR TITLE
Fix #43: Implement composite matcher

### LINKED ISSUE #43: Check a file-finder use-case
Use-case from @roxblnfk.

File-finder input (aliases do not matter):

```
@app/database/schema.php
@app/modules/*/database/schema.php
vendor/yii-megasoft/admin/database/schema/**.php
```

Expected output:

```
src/database/schema.php
src/modules/user/database/schema.php
src/modules/blog/database/schema.php
vendor/yii-megasoft/admin/database/schema/rbac/roles.php
vendor/yii-megasoft/admin/database/schema/rbac/assignes.php
vendor/yii-megasoft/admin/database/schema/menu.php
```

In case `src/database/schema.php` does not exist, either myself (by having a list of output files), or finder should throw and exception. The rest of the files matching are result of scanning of file system. If there are no such files, it doesn't matter. i.e. strict match in case the pattern isn't used, optional match if pattern is used.

### PR DESCRIPTION
**Pull request by samdark**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ❌
| Fixed issues  | #43

### CODE REVIEW

File: src/PathMatcher/CompositeMatcher.php [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+final class CompositeMatcher implements PathMatcherInterface
+{
+    private bool $matchAny;
+
+    /**
+     * @var array PathMatcherInterface[]
+     */
+    private array $matchers;
+
+    private function __construct(bool $matchAny, array $matchers)
```
Comment (viktorprogger) [inline src/PathMatcher/CompositeMatcher.php:23]:
```suggestion
    private function __construct(bool $matchAny, PathMatcherInterface ...$matchers)
```
I think typed variadic is more convenient way to check list items type

File: src/PathMatcher/CompositeMatcher.php [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+                $message = sprintf(
+                    'Matchers should contain instances of \Yiisoft\Files\PathMatcher\PathMatcherInterface, %s given.',
+                    $type
+                );
+                throw new InvalidArgumentException($message);
+            }
+        }
+        $this->matchers = $matchers;
+        $this->matchAny = $matchAny;
+    }
```
Comment (viktorprogger) [inline src/PathMatcher/CompositeMatcher.php:37]:
```suggestion
    private function __construct(bool $matchAny, PathMatcherInterface ...$matchers)
    {
        $this->matchers = $matchers;
        $this->matchAny = $matchAny;
    }
```
I think typed variadic is more convenient way to check list items type

File: src/PathMatcher/CompositeMatcher.php [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+/**
+ * Composite matcher allows combining several matchers.
+ */
+final class CompositeMatcher implements PathMatcherInterface
+{
+    private bool $matchAny;
+
+    /**
+     * @var array PathMatcherInterface[]
```
Comment (vjik) [inline src/PathMatcher/CompositeMatcher.php:15]:
```suggestion
     * @var PathMatcherInterface[]
```

File: src/PathMatcher/CompositeMatcher.php [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+    public static function all(PathMatcherInterface ...$matchers): self
+    {
+        return new self(false, ...$matchers);
+    }
+
+    public function match(string $path): ?bool
+    {
+        $allNulls = true;
+
+        /** @var PathMatcherInterface $matcher */
```
Comment (vjik) [inline src/PathMatcher/CompositeMatcher.php:53]:
```suggestion
```

**Review by rustamwin — APPROVED** [review]

**Pull request merged**